include/platform/mir/renderer/sw/pixel_source.h

```diff
@@ -30,6 +35,74 @@ namespace renderer
 namespace software
 {
 
+class BufferDescriptor
+{
+public:
+    virtual ~BufferDescriptor() = default;
+
+    virtual MirPixelFormat format() const = 0;
+    virtual geometry::Stride stride() const = 0;
+    virtual geometry::Size size() const = 0;
+};
+
+template<typename T>
+class Mapping : public BufferDescriptor
+{
+public:
+    /**
+     * Destroy the mapping of the buffer.
+     *
+     * This does not destroy the underlying buffer, only the CPU-accessible
+     * view of it.
+     */
+    virtual ~Mapping() = default;
+
+    virtual T* data() = 0;
+    virtual size_t len() const = 0;
+};
+
+/**
+ * A Buffer that can be mapped into CPU-accessible memory and directly written to.
+ */
+class WriteMappableBuffer
+{
+public:
+    virtual ~WriteMappableBuffer() = default;
+
+    virtual std::unique_ptr<Mapping<unsigned char>> map_writeable() = 0;
```

Comment:
Aaargh.

As soon as I try to document this I realise we need a different API. This is close, but you can't express performance-useful code like “only write to the bits of the buffer you care about”, or “write only to the damaged areas”.

On the other hand, how much do we care about performance here? :grimacing:

Reply:
Is the different API incompatible? Could it be added if we find we need it?

Having something that works now seems better than something perfect at some indefinite future moment.

Reply:
The different API would be incompatible, yes. Or, rather, it wouldn't change any symbol names, but it would render existing usages incorrect - with incorrect usages resulting in rendering from uninitialised memory.

It turns out that none of our *existing* code cares about this at all, and could live with a simpler “alloc-and-fill-a-buffer” API that doesn't have these questions. The decoration code particularly would benefit from an extended API, but this'll do for now.